x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationDenialMessages.java

```diff
@@ -92,7 +92,6 @@ static String remoteActionDenied(
         String action,
         String clusterAlias
     ) {
-        assert isIndexAction(action);
```

Comment:
I think we can apply the diff below instead. The advantage is that we get a slightly clearer and more consistent exception message:

Without the diff, the exception message is `action [cluster:admin/scripts/painless/execute] towards remote cluster [my_remote_cluster] is unauthorized for user [remote_search_user] with effective roles [remote_search] because no remote indices privileges apply for the target cluster`.

It's confusing that we have a cluster action, but say "no remote indices privileges".

With the diff, we get the "correct" index action name.

```diff
diff --git a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationDenialMessages.java b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationDenialMessages.java
index 82cea17cc6f..9d66318a1f0 100644
--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationDenialMessages.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationDenialMessages.java
@@ -92,6 +92,7 @@ class AuthorizationDenialMessages {
         String action,
         String clusterAlias
     ) {
+        assert isIndexAction(action);
         String userText = successfulAuthenticationDescription(authentication, authorizationInfo);
         String remoteClusterText = remoteClusterText(clusterAlias);
         return actionIsUnauthorizedMessage(action, remoteClusterText, userText)
diff --git a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java
index 162cabf5297..ca08f63a09b 100644
--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java
@@ -47,6 +47,7 @@ import org.elasticsearch.xpack.core.security.user.SystemUser;
 import org.elasticsearch.xpack.core.security.user.User;
 import org.elasticsearch.xpack.core.ssl.SSLService;
 import org.elasticsearch.xpack.security.Security;
+import org.elasticsearch.xpack.security.action.SecurityActionMapper;
 import org.elasticsearch.xpack.security.audit.AuditUtil;
 import org.elasticsearch.xpack.security.authc.ApiKeyService;
 import org.elasticsearch.xpack.security.authc.AuthenticationService;
@@ -385,7 +386,11 @@ public class SecurityServerTransportInterceptor implements TransportInterceptor
                                 )
                             );
                             if (roleDescriptorsIntersection.isEmpty()) {
-                                throw authzService.remoteActionDenied(authentication, action, remoteClusterAlias);
+                                throw authzService.remoteActionDenied(
+                                    authentication,
+                                    SecurityActionMapper.action(action, request),
+                                    remoteClusterAlias
+                                );
                             }
                             final var crossClusterAccessHeaders = new CrossClusterAccessHeaders(
                                 remoteClusterCredentials.credentials(),
```

Slight preference to keep the assertion around for now.

Reply:
> It's confusing that we have a cluster action, but say "no remote indices privileges".

This is a valid point. Using `SecurityActionMapper.action` to map `cluster:admin/scripts/painless/execute` to `indices:data/read/scripts/painless/execute` action for the access denied message on query cluster is more consistent with what would be reported by the fulfilling cluster anyway.

Thanks for looking into this!

Reply:
Patch applied in latest commit.

Reply:
@quux00 thanks, looks good from my end!